crates/task-manager/src/response_calculator.rs

```diff
@@ -0,0 +1,120 @@
+use rand::Rng;
+use std::{future::Future, marker::PhantomData};
+use tracing::info;
+
+use crate::TaskManagerError;
+
+/// Trait used to compute the response of a task.
+pub trait ResponseCalculator<Input, Output> {
+    /// Compute the response of a task.
+    ///
+    /// # Arguments
+    ///
+    /// * `task_index` - The index of the task.
+    /// * `input` - The input of the task.
+    ///
+    /// # Returns
+    ///
+    /// * `Result<Output, TaskManagerError>` - The response of the task.
+    fn compute_response(
+        &self,
+        task_index: u32,
+        input: Input,
+    ) -> impl Future<Output = Result<Output, TaskManagerError>>;
+}
+
+/// Implementation of the [`ResponseCalculator`] trait that uses a function to compute the response.
+#[derive(Debug)]
+pub struct FunctionResponseCalculator<Input, Output, F>
+where
+    F: AsyncFn(u32, Input) -> Result<Output, TaskManagerError>,
+{
+    /// The function used to compute the response.
+    pub compute_fn: F,
+    _input: PhantomData<Input>,
+    _output: PhantomData<Output>,
+}
+
+impl<Input, Output, F> FunctionResponseCalculator<Input, Output, F>
+where
+    F: AsyncFn(u32, Input) -> Result<Output, TaskManagerError>,
+{
+    /// Create a new `FunctionResponseCalculator`.
+    ///
+    /// # Arguments
+    ///
+    /// * `compute_fn` - The function used to compute the response.
+    ///
+    /// # Returns
+    ///
+    /// * `FunctionResponseCalculator<Input, Output, F>` - A new `FunctionResponseCalculator`.
+    pub fn new(compute_fn: F) -> Self {
+        Self {
+            compute_fn,
+            _input: PhantomData,
+            _output: PhantomData,
+        }
+    }
```

Comment:
This can then be:

```suggestion
    pub fn new<Input, Output, CF>(compute_fn: CF) -> Self
    where
        CF: Fn(u32, Input) -> Result<Output, TaskManagerError>,
    {
        Self(|a, b| async { compute_fn(a, b) })
    }
```

And we can add a second new for async functions:

```suggestion
    pub fn new_async<Input, Output, CF>(compute_fn: CF) -> Self
    where
        CF: AsyncFn(u32, Input) -> Result<Output, TaskManagerError>,
    {
        Self(compute_fn)
    }
```